cpp/src/ns/ns_evolution.cpp

```diff
@@ -471,6 +499,57 @@ double FreddiNeutronStarEvolution::R_Alfven() const {
 	return ns_str_->args_ns.R_Alfven(GM(), Mdot_in());
 }
 
+double FreddiNeutronStarEvolution::R_Alfven_basic() const {
+	return ns_str_->args_ns.R_Alfven_basic(GM(), Mdot_in());
+}//надо переписать это так же, как mu_magn
+
+
+
+double FreddiNeutronStarEvolution::R_Magn_bozzo18() const {
+	const double chi_oblique_rad = double(chioblique()) * M_PI / 180.;
+    const double alpha = 0.5; // как его взять честно??? его надо откуда-то вызвать
+	//eta (Bozzo2018) = 0.2, 
+	const double parametr_bozzo = 2. * m::pow<2>(0.2) / alpha; // на самом деле делить на alpha, как это сделать?
+	const double RA = R_Alfven_basic();
+    const double RCorrot = R_cor();
+	const double parametr_thetta = RCorrot  / RA;
+
+    //const vecd& H = Height();//почему так нельзя?
+    const double H_to_R_temp = h2rbozzo(); //временно
+	const bool sign_at_inf = 1; //sign on unfunuty should be minus
+
+    
+    const double xbozzo_0 = R()[first()] / RCorrot;
+    const double fxbozzo_0 = parametr_bozzo * ( (1. - std::pow(xbozzo_0, 1.5) ) * m::pow<2>(std::cos(chi_oblique_rad)) + H_to_R_temp * (8. - 5. * std::pow(xbozzo_0, 1.5)) * m::pow<2>(std::sin(chi_oblique_rad))) - std::pow(xbozzo_0 * parametr_thetta, 3.5);
+    
+    if (std::signbit(fxbozzo_0) == sign_at_inf){
+        return 0. ; 
+    }
+    
+	size_t jj;
+    double xbozzo;
+    double fxbozzo;
+    //double fxbozzo_prev;
+    //now compare the sign of expression with its sign at the infinity, break when sign changes
+    
+	for(jj = first(); jj <= last() - 2 ; jj++){
+        //double fxbozzo_prev = fxbozzo;
+		xbozzo = R()[jj] / RA;
+		fxbozzo = parametr_bozzo *( (1. - std::pow(xbozzo / parametr_thetta, 1.5) ) * m::pow<2>(std::cos(chi_oblique_rad)) + H_to_R_temp * (8. - 5. * std::pow(xbozzo / parametr_thetta, 1.5)) * m::pow<2>(std::sin(chi_oblique_rad))) - std::pow(xbozzo, 3.5);
+        if (std::signbit(fxbozzo) == sign_at_inf){
+            break;
+        }
+	}
+	if (jj == last() - 2){
+        std::cout<<"cannot find a root for R magnetosphere which is kinda weird\n";
```

Comment:
Printing something into standard input is not enough here. If is a logical error you should raise an exception

Reply:
Probably you should raise `RadiusCollapseException()`

Reply:
done